Fix #13088: Add pre-push checks for MyPy

hardikkat24 opened the pull request:
## Overview
<!--
READ ME FIRST:
Please answer *both* questions below and check off every point from the Essential Checklist!
If there is no corresponding issue number, fill in N/A where it says [fill_in_number_here] below in 1.
-->

1. This PR fixes or fixes part of #13088 .
2. This PR does the following: This PR adds pre-push checks for MyPy and adds skip-install flag in mypy script. This PR is stacked on PR#13054.

**Runtime data of mypy script:**
**Note:** This data assumes there is no mypy cache. But cache is created on the local system for mypy whenever mypy tests run.  This cache further speeds up the mypy checks. The speed reduced may vary from 2s-5s.
Files passed to mypy script: Files/directories that mypy checked
Number of files checked: Number of files checked by mypy
Number of Errors Thrown: Number of mypy errors
Time Taken: Time taken by mypy script to run with --skip-install flag enabled
| Files passed to mypy script | Number of files checked | Number of Errors Thrown | Time Taken (is sec)|
|---|---|---|---|
| core/controller/admin.py | 1 | 241 | 8.23 |
| core/domain/ | 225 | 23404 | 15.14 | 
| core/ | 465 | 38684 | 20.63 | 15 |

## Essential Checklist

- [x] The PR title starts with "Fix #bugnum: ", followed by a short, clear summary of the changes. (If this PR fixes part of an issue, prefix the title with "Fix part of #bugnum: ...".)
- [x] The linter/Karma presubmit checks have passed locally on your machine.
- [x] "Allow edits from maintainers" is checked. (See [here](https://help.github.com/en/github/collaborating-with-issues-and-pull-requests/allowing-changes-to-a-pull-request-branch-created-from-a-fork) for instructions on how to enable it.)
  - This lets reviewers restart your CircleCI tests for you.
- [x] The PR is made from a branch that's **not** called "develop".

## Proof that changes are correct

<!--
Add videos/screenshots of the user-facing interface to demonstrate that the changes made in this PR work correctly.
If the changes in your PRs are autogenerated via a script and you cannot provide proof for the changes then please leave a comment "No proof of changes needed because {{Reason}}".
-->

When Pre-push mypy checks pass:
![Screenshot from 2021-06-28 01-56-36](https://user-images.githubusercontent.com/54679643/123558579-a6f39b00-d7b4-11eb-9c0c-0e85e90dc0fd.png)

When Pre-push mypy checks fail:
![Screenshot from 2021-06-28 01-59-29](https://user-images.githubusercontent.com/54679643/123558615-e1f5ce80-d7b4-11eb-9878-ea0b2edc43fd.png)
![Screenshot from 2021-06-28 01-59-08](https://user-images.githubusercontent.com/54679643/123558619-e6ba8280-d7b4-11eb-8836-fd3518bac3f8.png)


## PR Pointers

- Make sure to follow the [instructions for making a code change](https://github.com/oppia/oppia/wiki/Contributing-code-to-Oppia#instructions-for-making-a-code-change).
- Oppiabot will notify you when you don't add a PR_CHANGELOG label. If you are unable to do so, please @-mention a code owner (who will be in the Reviewers list), or ask on [Gitter](https://gitter.im/oppia/oppia-chat).
- For what code owners will expect, see the [Code Owner's wiki page](https://github.com/oppia/oppia/wiki/Oppia%27s-code-owners-and-checks-to-be-carried-out-by-developers).
- Make sure your PR follows conventions in the [style guide](https://github.com/oppia/oppia/wiki/Coding-style-guide), otherwise this will lead to review delays.
- Never force push. If you do, your PR will be closed.
- Oppiabot can assign anyone for review/help if you leave a comment like the following: "{{Question/comment}} @{{reviewer_username}} PTAL"
- Some of the e2e tests are flaky, and can fail for reasons unrelated to your PR. We are working on fixing this, but in the meantime, if you need to restart the tests, please check the ["If your build fails" wiki page](https://github.com/oppia/oppia/wiki/If-your-build-fails).

hardikkat24 committed:
added files and added types to utils.py

hardikkat24 committed:
bug

hardikkat24 committed:
utils_test.py annotated

hardikkat24 committed:
changes

hardikkat24 committed:
workflow

hardikkat24 committed:
Tests and script changed

hardikkat24 committed:
work on backend tests and lint

hardikkat24 committed:
added workflow

hardikkat24 committed:
lint errors fixed

hardikkat24 committed:
lint errors fixed

hardikkat24 committed:
typing optional

hardikkat24 committed:
added test to shard

hardikkat24 committed:
workflow changed

hardikkat24 committed:
Merge branch 'develop' of https://github.com/oppia/oppia into m1.1

hardikkat24 committed:
workflow changed

hardikkat24 committed:
unicode bug resolved

hardikkat24 committed:
unicode errors

hardikkat24 committed:
error resolve

hardikkat24 committed:
workflow improved

hardikkat24 committed:
workflow changes

hardikkat24 committed:
workflow changed

hardikkat24 committed:
workflow changed

hardikkat24 committed:
workflow for 0.812

hardikkat24 committed:
lint issues fixed

hardikkat24 committed:
mypy_requirements.txt added and workflow modified

hardikkat24 committed:
mypy_req removed

hardikkat24 committed:
mypy_requirements added to codeowners

hardikkat24 committed:
increase test coverage

hardikkat24 committed:
added todos

hardikkat24 committed:
lint issues fixed by linebreaks

hardikkat24 committed:
lint issues fixed

hardikkat24 committed:
Merge branch 'develop' of https://github.com/oppia/oppia into m1.1

hardikkat24 committed:
typo corrected

hardikkat24 committed:
pre_push_hook and tests added

hardikkat24 committed:
test

hardikkat24 committed:
test2

hardikkat24 committed:
test3

hardikkat24 committed:
test4

hardikkat24 committed:
typing added to prerequisites list

hardikkat24 committed:
comment removed

hardikkat24 committed:
import typing test added

hardikkat24 committed:
utils.pt changed

hardikkat24 committed:
lint check script_test

hardikkat24 committed:
t1

hardikkat24 committed:
t2

hardikkat24 committed:
t3

hardikkat24 committed:
t4

hardikkat24 committed:
t5

hardikkat24 committed:
prerequisites mypy install

hardikkat24 committed:
changes

hardikkat24 committed:
files added

hardikkat24 committed:
annotations fixed

hardikkat24 committed:
lint issues

hardikkat24 committed:
Merge branch 'develop' of https://github.com/oppia/oppia into m1.1

hardikkat24 committed:
changes

hardikkat24 committed:
Merge branch 'm1.1' into m1.3

hardikkat24 committed:
self removed from argument

hardikkat24 committed:
changes

hardikkat24 committed:
test broken into 5

hardikkat24 committed:
test broken into 5-docum

hardikkat24 committed:
test added

hardikkat24 committed:
Merge branch 'm1.1' into m1.3

hardikkat24 committed:
optional removed and strict=true

hardikkat24 committed:
Merge branch 'develop' of https://github.com/oppia/oppia into m1.1

hardikkat24 committed:
denylist extended

hardikkat24 committed:
lint issues fixed

hardikkat24 committed:
link added

hardikkat24 committed:
changed test name

hardikkat24 committed:
parser desc changed

hardikkat24 committed:
changes

hardikkat24 committed:
script tests changed

hardikkat24 committed:
lint issues fixed

hardikkat24 committed:
lint issues fixed + codeowner change

hardikkat24 committed:
changes

hardikkat24 committed:
changes

hardikkat24 committed:
fixed merge conflicts

hardikkat24 committed:
run_mypy_checks

hardikkat24 committed:
actions merge

hardikkat24 committed:
Merge branch 'develop' of https://github.com/oppia/oppia into m1.1

hardikkat24 committed:
third_party removed

sajalasati commented:
@hardikkat24 Can you mention the runtime info of running mypy locally as a pre-push for changed files vs running it for all the files in the codebase in the PR description? Please provide helpful links and explanations along with that so that it can be understood what those stats actually mean.

hardikkat24 committed:
third_party in mypy checks

hardikkat24 committed:
fix tests failing

hardikkat24 committed:
lint issues fixed

hardikkat24 committed:
mypy workflow

hardikkat24 committed:
Merge branch 'develop' of https://github.com/oppia/oppia into m1.1

hardikkat24 committed:
Merge branch 'develop' of https://github.com/oppia/oppia into m1.1

hardikkat24 committed:
fixed type annotation errors

hardikkat24 committed:
Merge branch 'm1.1' into m1.3

hardikkat24 committed:
python3 to python2

hardikkat24 committed:
installation at top of script

hardikkat24 committed:
lint issues

hardikkat24 commented:
@sajalasati Now after adding install_third_party_libs to the script, you can see that the time taken by the script to run is 33s. What fixes do you suggest?

hardikkat24 committed:
Merge branch 'm1.1' into m1.3

hardikkat24 committed:
Merge branch 'develop' of https://github.com/oppia/oppia into m1.3

hardikkat24 committed:
skip_install flag introduced

hardikkat24 committed:
skip-install in workflow and pre-push hook

hardikkat24 commented:
@sajalasati PTAL.

sajalasati commented on scripts/run_mypy_checks_test.py:
```diff
@@ -111,34 +128,34 @@ def test_running_script_with_failure(self):
                 with self.assertRaisesRegexp(SystemExit, '1'):
                     run_mypy_checks.main(args=['--files', 'file1.py'])
 
     def test_main_failure(self):
```
Use a more descriptive name for this. What causes this failure? Mention that in the name as done in other tests.

hardikkat24 replied:
Done. PTAL

sajalasati commented on scripts/run_mypy_checks_test.py:
```diff
@@ -64,6 +74,13 @@ def mock_install_mypy_prerequisites_success():
             run_mypy_checks, 'install_mypy_prerequisites',
             mock_install_mypy_prerequisites_success)
 
+    def test_install_third_party_libraries_with_true(self):
```
With true what? Use a name like `test_install_third_party_with_skip_install_as_true` - or something around that. And why is there no assert statement here? What is it testing that you are testing and how you are doing that - that is not clear here.

hardikkat24 replied:
Changed the name. PTAL.
And the test is that running this method doesn't throw any error. We do this in our codebase. PTAL at the example [here](https://github.com/oppia/oppia/blob/4172bcc457b5fff70bb8ee70e596b0c7940b2fdb/scripts/run_e2e_tests_test.py#L145).

sajalasati replied:
cc @DubeySandeep - PTAL once!

sajalasati commented on scripts/run_mypy_checks_test.py:
```diff
@@ -64,6 +74,13 @@ def mock_install_mypy_prerequisites_success():
+    def test_install_third_party_libraries_with_true(self):
+        run_mypy_checks.install_third_party_libraries(True)
+
+    def test_install_third_party_libraries_with_false(self):
```
Ditto - Use a better test name here. And explain why an assert statement cannot be used here?

hardikkat24 replied:
Explained in the other example. Thanks!

sajalasati commented on scripts/run_mypy_checks.py:
```diff
@@ -794,7 +805,7 @@ def main(args=None):
     """Runs the MyPy type checks."""
     parsed_args = _PARSER.parse_args(args=args)
 
-    common.fix_third_party_imports()
```
@hardikkat24 Can you explain why this line is being removed here?

hardikkat24 replied:
Sorry, removed it by mistake.

sajalasati commented on scripts/pre_push_hook.py:
```diff
@@ -328,6 +329,14 @@ def start_linter(files):
+def start_mypy_checks():
+    """Starts the mypy type checks."""
+    task = subprocess.Popen(
+        [PYTHON_CMD, '-m', MYPY_TYPE_CHECK_MODULE, '--skip-install'])
```
@DubeySandeep Is it possible to do the same for linters too - i.e. include a skip install flag in the commands list itself, and do the installing third party libs part separately in main (or wherever correct). I thought this would be helpful as we wouldn't have to depend on linter run to ensure all third party libs are installed for mypy before it can run.
@hardikkat24 Did you explore this?

hardikkat24 replied:
@sajalasati In pre-push hook before mypy checks, we check for backend python lib inconsistencies. See [here](https://github.com/oppia/oppia/blob/4172bcc457b5fff70bb8ee70e596b0c7940b2fdb/scripts/pre_push_hook.py#L499)
So do you think that is enough?
And after that if there are files to lint, then linter is run with installs all backend python libs.
Thanks!

sajalasati replied:
> @sajalasati In pre-push hook before mypy checks, we check for backend python lib inconsistencies. See [here](https://github.com/oppia/oppia/blob/4172bcc457b5fff70bb8ee70e596b0c7940b2fdb/scripts/pre_push_hook.py#L499)
> So do you think that is enough?
> And after that if there are files to lint, then linter is run with installs all backend python libs.
> Thanks!

@hardikkat24 As per the function documentation, I don't think it's installing any dependencies, it's just checking for inconsistencies.

hardikkat24 replied:
@sajalasati I think that if the backend libraries are not installed fully, the script will stop executing due to inconsistencies. Am I missing something here?

sajalasati reviewed (changes requested):
@hardikkat24 Thanks! Left few comments. Can you please add a screenshot in PR description showing how does the error msg looks like for the user? Also include the time it took locally to run mypy on those files. And fix the failing type checks.

hardikkat24 committed:
add common.fix_third_party_imports()

hardikkat24 committed:
fix import dev_appserver error

hardikkat24 commented:
@sajalasati I have added data of time taken to run mypy checks depending on the files chosen. Thanks!

hardikkat24 committed:
change test names

hardikkat24 committed:
change test names-2

hardikkat24 commented:
@sajalasati The PR for failing type checks is already made. So I'm not adding them here. I have added the screenshots of pre-push checks and runtime data in the PR description. I have also addressed all your comments. PTAL.

hardikkat24 committed:
changes

sajalasati commented on scripts/run_mypy_checks_test.py:
```diff
@@ -83,62 +100,62 @@ def test_get_mypy_cmd_with_files(self):
                     run_mypy_checks.main(args=['--files', 'file1.py'])
 
-    def test_main_failure(self):
+    def test_main_failure_with_mypy_errors(self):
```
`test_main_failure_due_to_mypy_errors `

sajalasati reviewed (changes requested):
Thanks @hardikkat24! Looks good to me in general, @DubeySandeep @krishnarao22 PTAL at codeowner files? 
(@DubeySandeep there are 2 active threads I want you to look at - [this](https://github.com/oppia/oppia/pull/13114#discussion_r659327312) and [this](https://github.com/oppia/oppia/pull/13114#discussion_r659751847) - I won't block PR for this, just wanted you to take a look)

hardikkat24 committed:
test name changed

krishnarao22 reviewed (approved):
LGTM for 1 codeowner file.

sajalasati reviewed (approved):
LGTM!

DubeySandeep commented on scripts/pre_push_hook.py:
```diff
@@ -509,6 +518,13 @@ def main(args=None):
                         'Push failed, please correct the linting issues above.')
                     sys.exit(1)
 
+            mypy_check_status = start_mypy_checks()
```
Is it possible to run mypy checks only for the changed files? (similar to lint checks)

DubeySandeep replied:
This is important as the time for type check is expected to increase rapidly, right? (And we are trying to reduce the pre-push load as much as possible)

hardikkat24 replied:
@DubeySandeep The time to run type checks won't be much as you can see in the PR description. This time further reduces as MyPy creates cache. 
In linters, we can run tests on all the files. In MyPy tests we have a list of files and directories which do not have type annotations so we would have to check every changed file whether it exists in the list or not and type check only type annotated + changed file.
Do you think it is worth it? Thanks!

hardikkat24 replied:
And there was one more important point, a change somewhere in a file can result in mypy errors in other files. So I would say we should type check all the type annotated files where they are changed or unchanged.

hardikkat24 replied:
@DubeySandeep Let's suppose there are 2 files X and Y. X has function f definition and Y calls the function f.
Now in one PR, type annotation to file Y is added. Here we have added `# type: ignore[no-untyped-call]` after the function f call to tell mypy to not throw the error due to function f not having type annotation.
Now in a different PR, type annotations to file X is added. This will lead to mypy error in file Y. The error will be because the `#type: ignore[no-untyped-call]` we added earlier is not used now as function f has type annotations.
Now there are two cases:
1. Running pre-push type checks on all files: The error in file Y will be thrown and developer will fix them before pushing.
2. Running pre-push check on changed files: The error in file Y will not be thrown if file Y is not changed in the PR. The developer will see this error in CI checks and will have to fix this and push again.

So I would highly suggest running pre-push type checks on all the type annotated files.

sajalasati replied:
@hardikkat24 Please edit the above msg and also explain with the example you discussed with me - how was mypy handling that error before (in PR for Y file) and what error would be shown had we been running mypy check locally for all files. (An example would help in understanding your argument better)

hardikkat24 replied:
@sajalasati Added the example.

DubeySandeep replied:
@hardikkat24 I'm not sure why pre-push check should rely on different PRs?

DubeySandeep replied:
As per our offline discussion, we're going to run checks for the entire codebase. Reasons:

1. Mypy cache for files are operated with file hash so the caches are going to be reliable and useful.
2. If a commit changes function definition then other modules calling such functions with invalid params won't get caught!

DubeySandeep reviewed (changes requested):
@hardikkat24 I've left a comment, PTAL!

hardikkat24 commented:
@DubeySandeep PTAL. Explained the approach!

DubeySandeep commented on scripts/pre_push_hook.py:
```diff
@@ -328,6 +329,14 @@ def start_linter(files):
     return task.returncode
 
 
+def start_mypy_checks():
```
```suggestion
def execute_mypy_checks():
```

hardikkat24 replied:
Done

DubeySandeep commented on scripts/pre_push_hook.py:
```diff
@@ -328,6 +329,14 @@ def start_linter(files):
 
 
+def start_mypy_checks():
+    """Starts the mypy type checks."""
```
Add docstring to Returns*

hardikkat24 replied:
Done

DubeySandeep commented on scripts/pre_push_hook.py:
```diff
@@ -509,6 +518,13 @@ def main(args=None):
+                python_utils.PRINT(
+                    'Push failed, please correct the mypy type annotation '
+                    'issues above.')
+                sys.exit(1)
```
```suggestion
                sys.exit(mypy_check_status)
```

hardikkat24 replied:
Done

DubeySandeep commented on scripts/run_mypy_checks.py:
```diff
@@ -28,9 +28,6 @@
-# libraries that use the builtins python module (e.g. build, python_utils).
-install_third_party_libs.main()
 
 import python_utils # isort:skip  pylint: disable=wrong-import-position, wrong-import-order
```
Can we remove the flags now?

hardikkat24 replied:
Yes. Removed.

DubeySandeep reviewed (approved):
@hardikkat24 LGTM, I've left a few minor comments, PTAL!

DubeySandeep reviewed (changes requested):
[Marking "request changes" to avoid accidental merge!]

hardikkat24 committed:
changes

hardikkat24 committed:
Merge branch 'develop' of https://github.com/oppia/oppia into m1.3

hardikkat24 commented:
@DubeySandeep PTAL.
